sympy/printing/printer.py

```diff
@@ -55,135 +61,115 @@
         |
         `-- p._print_Basic(expr)
 
-    if ._print_Rational method exists in the printer, then it is called,
-    and the result is returned back.
+    if ``._print_Rational`` method exists in the printer, then it is called,
+    and the result is returned back. Otherwise, the printer tries to call
+    ``._print_Number`` and so on.
 
-    otherwise, we proceed with trying Rational bases in the inheritance
-    order.
-
-3. As fall-back use the emptyPrinter method for the printer.
-
-    As fall-back self.emptyPrinter will be called with the expression. If
-    not defined in the Printer subclass this will be the same as str(expr).
-"""
+3. **As a fall-back use the emptyPrinter method for the printer.**
 
-from __future__ import print_function, division
+    As fall-back ``self.emptyPrinter`` will be called with the expression. If
+    not defined in the Printer subclass this will be the same as ``str(expr)``.
 
-from sympy import Basic, Add
+Example of Custom Printer and Custom Printing Method
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-from sympy.core.core import BasicMeta
+.. _printer_example:
 
-from functools import cmp_to_key
+This example shows how you can define a custom Printer and a custom printing method
+for your custom SymPy object. See the code below::
 
+    from sympy import Basic, Function, Symbol
+    from sympy.printing.printer import Printer
+    from sympy.printing.latex import print_latex
+    from sympy.core.basic import Basic
 
-class Printer(object):
-    """Generic printer
+    class MyBasic(Basic):
+        \"\"\" Our custom SymPy object. You can also subclass sympy.core.Expr
+        or sympy.core.Function. Both of them are subclasses of Basic.
+        \"\"\"
 
-    Its job is to provide infrastructure for implementing new printers easily.
+        def __init__(self, text):
+            self.text = text
 
-    Basically, if you want to implement a printer, all you have to do is:
+        def __str__(self):
+            return self.text
 
-    1. Subclass Printer.
+        def _latex(self, printer=None):
+            # This is the printmethod of LatexPrinter.
+            return \"\\\\textbf{{{}}}\".format(self.text)
 
-    2. Define Printer.printmethod in your subclass.
-       If a object has a method with that name, this method will be used
-       for printing.
+    class MyPrinter(Printer):
```

Comment:
> The original example had _print_MyClass(self, expr) but MyClass wasn't defined anywhere so you could not compile the example. Therefore I also decided to add the code for MyClass and I renamed MyClass to MyBasic because it subclasses sympy.core.basic.

I think the subclassed printer should override an existing SymPy function. It doesn't really make sense to subclass a printer for a custom class, because you can just define the printer on the class. So in this example, it is a little unclear why you would even want to subclass a printer, and why we have both ways. But generally the recommendations are:

- If you have a custom SymPy class, add the printmethod to that class. 
- If you want to change the printing for an existing SymPy class, subclass the appropriate printer(s).

> By "Usually you want to subclass an existing printer, like StrPrinter or LatexPrinter." you mean that we also want to have StrPrinter in the example instead of Printer or you just want me to add a comment on this thing. I don't see a reason why directly subclassing Printer is a bad idea.

Again, it's not a realistic example. The real use-case is to change how the existing printers work. StrPrinter and LatexPrinter are the most common (also the pretty printer, but that also has some advanced stuff because it use to use the pretty form stuff). 

You can see lots of examples of things people want to do in [StackOverflow](https://stackoverflow.com/search?q=%5Bsympy%5D+printer).  [Here's](https://stackoverflow.com/questions/43350381/sympy-modifying-latex-output-of-derivatives/43356409#43356409) an example of the sort of thing that could make a nice example, becaues it's short, and it's also something that people actually would want to do.